pkgs/build-support/docker/default.nix

```diff
@@ -886,4 +888,70 @@ rec {
         done;
       '' + extraCommands;
     });
+
+  # This function builds a docker image that behaves like a nix-shell
+  buildNixShellImage =
+    # Use the environment of this derivation
+    { name
+    , drv
+    , tag ? null
+    # Extra packages available inside the image
+    , extraPackages ? []
+    # Extra bash commands to execute when the image is ran
+    , extraInit ? ""
+    }:
+    # Remove once https://github.com/NixOS/nixpkgs/pull/77952 is merged
+    let
+
+      finalDrv = drv.overrideAttrs (old: {
+        nativeBuildInputs = old.nativeBuildInputs or [] ++ extraPackages;
+      });
+
+      setupPackage = runCommandNoCC "setup" {
+        profile = ''
+          # This mimics part of nix-shell's initialization
+          source $stdenv/setup
+          set +e
+          eval "$shellHook"
+          cd /home
+        '' + extraInit;
+        passAsFile = [ "profile" ];
+      } ''
+        mkdir -p $out/{etc/ssl/certs,tmp,bin,usr/bin,home/{build,outputs}}
+        ln -s ${bashInteractive}/bin/sh $out/bin/sh
+        ln -s ${coreutils}/bin/env $out/usr/bin/env
+        cp "$profilePath" $out/etc/profile
+        ln -s "${cacert}/etc/ssl/certs/ca-bundle.crt" $out/etc/ssl/certs/ca-certificates.crt
+      '';
+    in buildLayeredImage {
+      inherit name tag;
+      contents = [ setupPackage ];
+      # These folders should be writable
+      extraCommands = "chmod +w -R tmp home";
+
+      # Don't run this image as root because many programs don't like that
+      config.User = "1000:1000";
+      # Set the owner of the files to that user as well
+      uid = 1000;
+      gid = 1000;
+
+      # Pass --login such that /etc/profile gets loaded
+      # Use bashInteractive for tab completion support and such
+      config.Entrypoint = [ "${bashInteractive}/bin/bash" "--login" ];
```

Comment:
This is different from the `bash` image, which has this entrypoint https://github.com/tianon/docker-bash/blob/master/5.1/docker-entrypoint.sh

When I try this in a normal session I get
```
$ bash --login echo foo
/run/current-system/sw/bin/echo: /run/current-system/sw/bin/echo: cannot execute binary file
```

Which seems to be a cryptic way of telling us that it's refusing to interpret an ELF executable as a script.

Passing a script does work
```
$ bash --login <(echo "echo hello") asdf
hello
```

So I think it should be like the official bash entrypoint.

The `Cmd` should probably be [`["bash"]`](https://github.com/tianon/docker-bash/blob/667f5e52f50ef8cdf3d03ab0aa86b9f90c77c726/5.1/Dockerfile#L127).

We'll need a test for this. You can add cases to `nixos/tests/docker-tools.nix`, based on a new example you can add to `pkgs/build-support/docker/examples.nix`.

Reply:
Not sure I get the motivation. `nix-shell` is bash-based, but it's not an image for the purpose of executing bash. In this case, we need some way to source stdenv and execute shellHook when it starts, which I'm now doing by passing `--login` and writing the shell init to `/etc/profile`. Currently this allows you to either do `docker run -it <image>`, then run `genericBuild`, or you can also do
```
docker run -it <image> -c genericBuild
```

And this works, which I don't think would be the case with your suggestion.

Reply:
I agree we seem to have different use cases in mind, but we can probably support both.

 - use the nix-shell image to build something
 - use the nix-shell image as a developer environment

For building, we should invoke the `builder` with `args` and not wrongly assume `bashInteractive`.
For interactive use, I do think we should emulate the official bash image behavior.

Can't we support both?

Reply:
I now changed it to have no entrypoint at all, however there's bash, all of coreutils and a special `buildDerivation` binary in the `PATH`. In addition, the `Cmd` is set to `[ "bash" ]`. I managed to get around the `--login` requirement by using `~/.bashrc` instead. This means by default you'll get a `nix-shell` environment, but you can also do any of
```
$ docker run -it <image> # nix-shell
$ docker run -it <image> bash # nix-shell
$ docker run -it <image> buildDerivation # build the derivation directly
$ docker run -it <image> bash -c 'buildDerivation && ls $out/bin' # builds derivation and lists the binaries it produced